rclcpp/src/rclcpp/utilities.cpp

```diff
@@ -42,116 +43,233 @@
+      if (msg != error_string) {
+        strncpy(error_string, msg, error_length);
+        msg[error_length - 1] = '\0';
+      }
 #else
-    int error_status = strerror_r(errno, error_string, error_length);
-    if (error_status != 0) {
-      throw std::runtime_error("Failed to get error string for errno: " + std::to_string(errno));
-    }
+      int error_status = strerror_r(errno, error_string, error_length);
+      if (error_status != 0) {
+        throw std::runtime_error("Failed to get error string for errno: " + std::to_string(errno));
+      }
 #endif
 #else
-    strerror_s(error_string, error_length, errno);
+      strerror_s(error_string, error_length, errno);
 #endif
-    // *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
-    throw std::runtime_error(
-      std::string("Failed to set SIGINT signal handler: (" + std::to_string(errno) + ")") +
-      error_string);
-    // *INDENT-ON*
-  }
+      // *INDENT-OFF* (prevent uncrustify from making unnecessary indents here)
+      throw std::runtime_error(
+        std::string("Failed to set SIGINT signal handler: (" + std::to_string(errno) + ")") +
+        error_string);
+      // *INDENT-ON*
+    }
 
 #ifdef HAS_SIGACTION
-  return old_action;
+    return old_action;
 #else
-  return old_signal_handler;
+    return old_signal_handler;
 #endif
-}
+  }
 
-void
+  static
+  void
 #ifdef HAS_SIGACTION
-signal_handler(int signal_value, siginfo_t * siginfo, void * context)
+  signal_handler(int signal_value, siginfo_t * siginfo, void * context)
 #else
-signal_handler(int signal_value)
+  signal_handler(int signal_value)
 #endif
-{
-  RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "signal_handler(signal_value=%d)", signal_value);
+  {
+    RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "signal_handler(signal_value=%d)", signal_value);
 
 #ifdef HAS_SIGACTION
-  if (old_action.sa_flags & SA_SIGINFO) {
-    if (old_action.sa_sigaction != NULL) {
-      old_action.sa_sigaction(signal_value, siginfo, context);
-    }
-  } else {
-    if (
-      old_action.sa_handler != NULL &&  // Is set
-      old_action.sa_handler != SIG_DFL &&  // Is not default
-      old_action.sa_handler != SIG_IGN)  // Is not ignored
-    {
-      old_action.sa_handler(signal_value);
+    if (old_action.sa_flags & SA_SIGINFO) {
+      if (old_action.sa_sigaction != NULL) {
+        old_action.sa_sigaction(signal_value, siginfo, context);
+      }
+    } else {
+      if (
+        old_action.sa_handler != NULL &&  // Is set
+        old_action.sa_handler != SIG_DFL &&  // Is not default
+        old_action.sa_handler != SIG_IGN)  // Is not ignored
+      {
+        old_action.sa_handler(signal_value);
+      }
     }
-  }
 #else
-  if (old_signal_handler) {
-    old_signal_handler(signal_value);
-  }
+    if (old_signal_handler) {
+      old_signal_handler(signal_value);
+    }
 #endif
+    signal_received.exchange(true);
+    events_condition_variable.notify_one();
```

Comment:
I think @hidmic mentioned this before, [calling  `notify_one()` from a signal handler is undefined behavior](https://stackoverflow.com/questions/31521182/c11-use-condition-variable-in-signal-handler/31521467). Is there anything else that can be used to wake the thread?

Reply:
I asked him about this, but I don't remember his answer. Do you know of an alternative? Only thing I can think of is a busy wait in the signal handler loop and an atomic, but that doesn't sound very nice.

Reply:
Searching only turns up things that are platform specific.

`sem_post()` for POSIX platforms http://pubs.opengroup.org/onlinepubs/009695399/functions/sem_post.html

I'm not sure about windows. I can't tell from the documentation if windows semaphores are signal safe, https://docs.microsoft.com/en-us/windows/desktop/sync/using-semaphore-objects

Reply:
@sloretz Yeah, that's the thing with signal handlers :sweat_smile: 

AFAIK there's no completely safe cross-platform synchronization mechanism that can be used from a signal handler (if only Windows was POSIX compliant...). Busy waits, with or without a timeout, are a portable option. Not a very neat one though. We can also do some more research on how to do this right for each platform.

I'd expect the current implementation to work in almost all cases though. Unless a signal comes right when the thread is about to do something with the condition variable. and that only takes place during a negligible fraction of the process lifetime.

Reply:
>Which I gather isn't safe enough? Is that your conclusion as well @sloretz?

It uses `pthread_cond_signal`, which [this page says](https://linux.die.net/man/3/pthread_cond_signal) 
```
It is not safe to use the pthread_cond_signal() function in a signal handler that is invoked 
asynchronously. Even if it were safe, there would still be a race between the test of the Boolean 
pthread_cond_wait() that could not be efficiently eliminated. 
```

Reply:
Ok, so I was wrong, this is not safe for pthread, see:

> It is not safe to use the pthread_cond_signal() function in a signal handler that is invoked asynchronously. Even if it were safe, there would still be a race between the test of the Boolean pthread_cond_wait() that could not be efficiently eliminated.
>
> Mutexes and condition variables are thus not suitable for releasing a waiting thread by signaling from code running in a signal handler.
> -- https://linux.die.net/man/3/pthread_cond_signal